lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -950,18 +938,122 @@ static Value createBarrierAlloc(scf::ForOp &forOp, unsigned distance) {
   return barrierAlloc;
 }
 
+struct AsyncLoad {
+  AsyncLoad(Operation *loadOp, Value alloc) : loadOp(loadOp), alloc(alloc) {}
+  Operation *loadOp;
+  Value alloc;
+  Value barrier;
+  Operation *waitOp = nullptr;
+  bool isTMALoad = false;
+};
+
+// Create barriers and wait ops for the async loads. Barriers may be shared by
+// multiple loads is the schedule allows it.
+static void createTMABarrierAndWait(
+    scf::ForOp &forOp, SmallVector<AsyncLoad> &asyncLoads, Value insertIdx,
+    Value extractIdx, Value phase, int numBuffers, CoarseSchedule &schedule,
+    SmallVector<Value> &barriers,
+    const llvm::MapVector<Operation *, LoadInfo> &loadToInfo) {
+  llvm::SmallDenseMap<Operation *, AsyncLoad *> loadToAsyncLoad;
+  for (AsyncLoad &asyncLoad : asyncLoads) {
+    loadToAsyncLoad[asyncLoad.loadOp] = &asyncLoad;
+  }
+  SmallVector<SmallVector<AsyncLoad *>> loadGroups;
+  llvm::SmallDenseSet<Operation *> visited;
+  Block *forBlock = forOp.getBody();
+  // Find groups of loads that can share the same barrier. We look consecutive
+  // loads and check that there are uses in between.
+  for (AsyncLoad &asyncLoad : asyncLoads) {
+    if (!asyncLoad.isTMALoad || visited.count(asyncLoad.loadOp))
+      continue;
+    llvm::SmallDenseSet<Operation *> users;
+    SmallVector<AsyncLoad *> group;
+    auto addToGroup = [&](AsyncLoad *loadInfo) {
+      group.push_back(loadInfo);
+      visited.insert(loadInfo->loadOp);
+      for (Operation *user : loadInfo->loadOp->getUsers()) {
+        auto it = loadToInfo.find(loadInfo->loadOp);
+        if (it != loadToInfo.end()) {
+          // Special case for MMAv3 loads, we can ignore the alloc and only
+          // consider uses of the alloc op since it will be removed.
+          if (it->second.loadIsMMAV3) {
+            auto alloc = cast<ttg::LocalAllocOp>(
+                (*loadInfo->loadOp->getUsers().begin()));
+            if (alloc->getBlock() == forBlock) {
+              users.insert(alloc->getUsers().begin(), alloc->getUsers().end());
+              continue;
+            }
+          }
+        }
+        Operation *userInBlock = forBlock->findAncestorOpInBlock(*user);
```

Comment:
Will this break if load is under an `if`? I don't think we are restricting loads to be only in the main block of the loop, right? Maybe we should.

Reply:
I think so, this is why I call `findAncestorOpInBlock` so that even if the op is nested it still work

Reply:
oh you mean whether it would still re-use the barrier? I guess the rest of the code would be conservative. We can relax it as needed. I was trying to go for something minimal